File: online-judges/leetcode/min-cost-climbing-stairs.rs
// https://leetcode.com/problems/min-cost-climbing-stairs/

use std::cmp::min;

impl Solution {
    pub fn min_cost_climbing_stairs(cost: Vec<i32>) -> i32 {
        let mut dp = vec![cost[0]; cost.len()];
        dp[1] = cost[1];

        for i in 2..cost.len() {
            dp[i] = cost[i] + min(dp[i - 1], dp[i - 2]);
        }

        min(dp[cost.len() - 1], dp[cost.len() - 2])
    }
}
//

struct Solution {}

#[test]
fn test1() {
    let cost = vec![10, 15, 20];
    let min_cost = Solution::min_cost_climbing_stairs(cost);
    assert!(min_cost == 15);
}

#[test]
fn test2() {
    let cost = vec![1, 100, 1, 1, 1, 100, 1, 1, 100, 1];
    let min_cost = Solution::min_cost_climbing_stairs(cost);
    assert!(min_cost == 6);
}

File: online-judges/leetcode/min-cost-climbing-stairs.py
# https://leetcode.com/problems/min-cost-climbing-stairs/

from typing import List


class Solution:
    def minCostClimbingStairs(self, cost: List[int]) -> int:
        cache = dict()

        def min_cost(cur):
            if cur in cache:
                return cache[cur]

            if cur >= len(cost):
                return 0

            cache[cur] = cost[cur] + min(min_cost(cur + 1), min_cost(cur + 2))
            return cache[cur]

        return min(min_cost(0), min_cost(1))


# Tests
solver = Solution()
assert solver.minCostClimbingStairs([10, 15, 20]) == 15
assert solver.minCostClimbingStairs([1, 100, 1, 1, 1, 100, 1, 1, 100, 1]) == 6
